### PR TITLE
Add option to monitor DNS with MTU-sized payload len

### PR DESCRIPTION
**Pull request by joestringer**
Allow entire DNS packets to be monitored for introspection into DNS conversations.

Question:
* Should this also be hidden behind a runtime flag?

Tested via minikube.

---

This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/5551)

### CODE REVIEW

**Comment by joestringer** [comment]
test-me-please

File: bpf/lib/common.h [hunk 1]
```diff
@@ -356,6 +356,11 @@ ipv6_ct_tuple_get_daddr(struct ipv6_ct_tuple *tuple)
 #endif
 }
 
+static inline bool ipv6_is_dns(struct ipv6_ct_tuple *tuple)
+{
+	return tuple->nexthdr == IPPROTO_UDP && tuple->dport == bpf_htons(53);
```
Comment (tgraf) [inline bpf/lib/common.h]: Is the sport/dport reversed here? I assume you want to match on the sport being 53 to catch the DNS response with the A and AAAA record.

**Review by tgraf — CHANGES_REQUESTED** [review]
Looks great overall, minor question on sport/dport.

File: bpf/lib/common.h (hunk 1 above)
Comment (joestringer) [reply bpf/lib/common.h]: Yes, it's reversed. I can add a comment to explain.

**Comment by joestringer** [comment]
@tgraf Updated the comment and also restricted to only increase the capture length for replies (previously it was capturing all requests/responses)

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
Too many instructions. Playing around with some techniques to fold the new checks together.

File: bpf/lib/common.h [hunk 2]
```diff
@@ -335,6 +335,14 @@ enum {
 	CT_RELATED,
 };
 
+/* filter_dns returns true if the packet is a DNS response, false otherwise. */
+static inline bool filter_dns(__u8 nexthdr, __u16 dport, __u8 dir)
+{
+	/* The dport represents the destination of the connection. */
+	return ((nexthdr == IPPROTO_UDP || nexthdr == IPPROTO_UDP) &&
```
Comment (tgraf) [inline bpf/lib/common.h]: This looks wrong:
`(nexthdr == IPPROTO_UDP || nexthdr == IPPROTO_UDP)` Did you mean to include TCP?

Comment (joestringer) [reply bpf/lib/common.h]: Fixed.

**Comment by joestringer** [comment]
test-me-please

**Review by tgraf — APPROVED** [review]

**Pull request merged**